### PR TITLE
Feature: Rename component id

### PR DESCRIPTION
**Pull request by justinehell**
Link to the issue https://github.com/ditrit/leto-modelizer/issues/340
:information_source:  Don't forget to modify the changelog.md before merging this branch.

### CODE REVIEW

File: src/models/ComponentAttribute.js [hunk 1]
```diff
@@ -84,6 +84,37 @@ class ComponentAttribute {
       || this.__ruleRegexValidation();
   }
 
+  /**
+   * Get the reference value of the attribute.
+   * @returns {(string | boolean | number | Array | ComponentAttribute[])}
+   * The value of the attribute.
+   */
+  getReferenceValue() {
```
Comment (k-y0u) [inline src/models/ComponentAttribute.js:92]: Is it not already  possible to access `this.value` without a function ? If not, is it not a better solution to use getter & setter instead, like above `isVariable?`

File: src/models/ComponentAttribute.js [hunk 2]
```diff
@@ -84,6 +84,37 @@ class ComponentAttribute {
       || this.__ruleRegexValidation();
   }
 
+  /**
+   * Get the reference value of the attribute.
+   * @returns {(string | boolean | number | Array | ComponentAttribute[])}
+   * The value of the attribute.
```
Comment (k-y0u) [inline src/models/ComponentAttribute.js:90]: The value of the reference.

File: src/models/ComponentAttribute.js [hunk 3]
```diff
@@ -84,6 +84,37 @@ class ComponentAttribute {
...
+    this.value = value;
+  }
+
+  /**
+   * Replace the old ID with the new ID in the link attribute's value.
+   * @param {string} oldId - The old ID.
+   * @param {string} newId - The new ID.
+   */
+  replaceLinkId(oldId, newId) {
+    const index = this.value.indexOf(oldId);
```
Comment (k-y0u) [inline src/models/ComponentAttribute.js:111]: If `this.value` is a boolean (cf. l.98), it should not work.

**Review by k-y0u — CHANGES_REQUESTED** [review]

File: src/models/ComponentAttribute.js (hunk 1 above)
Comment (Zorin95670) [reply src/models/ComponentAttribute.js:92]: We can access, but the `Reference` value need to be overriden by other plugin like Terraform

File: src/models/ComponentAttribute.js (hunk 3 above)
Comment (Zorin95670) [reply src/models/ComponentAttribute.js:111]: a link can't be a boolean

File: src/models/ComponentAttribute.js [hunk 4]
```diff
@@ -84,6 +84,37 @@ class ComponentAttribute {
...
+   * @returns {(string | boolean | number | Array | ComponentAttribute[])}
+   * The value of the attribute.
+   */
+  getReferenceValue() {
+    return this.value;
+  }
+
+  /**
+   * Set the reference value of the attribute.
+   * @param {(string | boolean | number | Array | ComponentAttribute[])} value
```
Comment (Zorin95670) [inline src/models/ComponentAttribute.js:98]: only string here

File: src/models/ComponentAttribute.js [hunk 5]
```diff
@@ -84,6 +84,37 @@ class ComponentAttribute {
       || this.__ruleRegexValidation();
   }
 
+  /**
+   * Get the reference value of the attribute.
+   * @returns {(string | boolean | number | Array | ComponentAttribute[])}
```
Comment (Zorin95670) [inline src/models/ComponentAttribute.js:89]: only string here

File: src/models/Component.js [hunk 6]
```diff
@@ -196,6 +204,37 @@ class Component extends FileInformation {
...
+  getAttributesByType(...types) {
+    const result = [];
+
+    this.__setAttributesByType(this.attributes, result, types);
+
+    return result;
+  }
+
+  /**
+   * Get attribute from attributes list corresponding to the given type.
```
Comment (Zorin95670) [inline src/models/Component.js:221]: set attributes corresponding to the given type in the provided array.

File: src/models/DefaultData.js [hunk 7]
```diff
@@ -118,6 +118,39 @@ class DefaultData {
     return this.components.find((component) => component.id === id) || null;
   }
 
+  /**
+   * Get all components ids.
+   * @returns {string[]} An array of ids.
+   */
+  getComponentsIds() {
```
Comment (Zorin95670) [inline src/models/DefaultData.js:125]: not needed

File: demo/src/App.vue [hunk 8]
```diff
@@ -53,6 +79,14 @@ function reset() {
   plugin.draw('root', readOnly.value);
 }
 
+function renameComponent() {
+  plugin.data.renameComponentId(selectedId.value, renamedId.value);
+  plugin.draw('root', readOnly.value);
+  componentsIds.value = plugin.data.getComponentsIds();
```
Comment (Zorin95670) [inline demo/src/App.vue:85]: use you .map here

File: cypress/e2e/rename-component-id.feature [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    Then I expect ".link" appear 8 times on screen
+    And  I expect "#link-server-laptop_link-server1-external1" is visible
+    And  I expect "#svg-external1" is visible
+
+    When I select "external1" in "select"
+    And  I set on "#rename-input" text "renamed"
+    And  I click on "#rename-component"
+    Then I expect ".link" appear 8 times on screen
+    And  I expect "#svg-renamed" is visible
+    And  I expect "#link-server-laptop_link-server1-renamed" is visible
```
Comment (Zorin95670) [inline cypress/e2e/rename-component-id.feature:18]: Check if this not exists before make the renaming

File: cypress/e2e/rename-component-id.feature [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+  Scenario: Rename component should automatically update all the Reference attributes targeting the renamed component.
+    Then I expect "#svg-network1" is visible
+    And I expect "#server1" is visible
+    And I expect "#server2" is visible
+
+    When I select "network1" in "select"
+    And  I set on "#rename-input" text "renamed"
+    And  I click on "#rename-component"
+    Then I expect "#svg-renamed" is visible
```
Comment (Zorin95670) [inline cypress/e2e/rename-component-id.feature:29]: Check if this not exists before make the renaming

**Review by Zorin95670 — CHANGES_REQUESTED** [review]

File: src/models/ComponentAttribute.js (hunk 2 above)
Comment (justinehell) [reply src/models/ComponentAttribute.js:90]: done

File: src/models/ComponentAttribute.js (hunk 4 above)
Comment (justinehell) [reply src/models/ComponentAttribute.js:98]: done

File: src/models/ComponentAttribute.js (hunk 5 above)
Comment (justinehell) [reply src/models/ComponentAttribute.js:89]: done

File: src/models/Component.js (hunk 6 above)
Comment (justinehell) [reply src/models/Component.js:221]: done

File: src/models/DefaultData.js (hunk 7 above)
Comment (justinehell) [reply src/models/DefaultData.js:125]: done

File: demo/src/App.vue (hunk 8 above)
Comment (justinehell) [reply demo/src/App.vue:85]: done

File: cypress/e2e/rename-component-id.feature (hunk 9 above)
Comment (justinehell) [reply cypress/e2e/rename-component-id.feature:18]: done

File: cypress/e2e/rename-component-id.feature (hunk 10 above)
Comment (justinehell) [reply cypress/e2e/rename-component-id.feature:29]: done

File: src/models/ComponentAttribute.js (hunk 3 above)
Comment (justinehell) [reply src/models/ComponentAttribute.js:151]: done

File: tests/unit/models/ComponentAttribute.spec.js [hunk 11]
```diff
@@ -101,6 +101,108 @@ describe('Test class: ComponentAttribute', () => {
...
+            type: 'Link',
+            rules: {
+              values: null,
+              min: 1,
+              max: 5,
+            },
+          },
+        });
+
+        componentAttribute.replaceLink('oldId', 'newId');
```
Comment (Zorin95670) [inline tests/unit/models/ComponentAttribute.spec.js:130]: Misplace test

File: tests/unit/models/ComponentAttribute.spec.js [hunk 12]
```diff
@@ -101,6 +101,108 @@ describe('Test class: ComponentAttribute', () => {
...
+
+    describe('Test method: replaceLink', () => {
+      it('should replace the old id with the new id', () => {
+        const componentAttribute = new ComponentAttribute({
+          value: 'value',
+        });
+
+        expect(componentAttribute.value).toEqual('value');
+
+        componentAttribute.setReferenceValue('newValue');
```
Comment (Zorin95670) [inline tests/unit/models/ComponentAttribute.spec.js:200]: Misplace test

File: src/models/Component.js [hunk 13]
```diff
@@ -196,6 +204,37 @@ class Component extends FileInformation {
...
+  /**
+   * Set attributes corresponding to the given type in the provided array.
+   * Search in sub-attributes of "Object" attributes also.
+   * @param {ComponentAttribute[]} attributes - Attributes list.
+   * @param {ComponentAttribute[]} result - The array to store the found attributes.
+   * @param {...string} types - One or more attribute types to search for.
+   * @private
+   */
+  __setAttributesByType(attributes, result, types) {
+    for (let index = 0; index < attributes.length; index += 1) { // NOSONAR
```
Comment (Zorin95670) [inline src/models/Component.js:229]:
```js
attributes.forEach((attribute) => {
  if (attribute?.type === 'Object') {
    this.__setAttributesByType(attribute.value, result, types);
  }
  
  if (types.includes(attribute?.definition.type)) {
    result.push(attribute);
  }
});
```

File: tests/unit/models/Component.spec.js [hunk 14]
```diff
@@ -564,6 +576,65 @@ describe('Test class: Component', () => {
...
+
+      it('Should return an array with matching attribute types', () => {
+        const result = component.getAttributesByType('Reference', 'Link');
+
+        expect(result).toHaveLength(2);
+        expect(result[0].definition.type).toBe('Reference');
+        expect(result[1].definition.type).toBe('Link');
+      });
+
+      it('Should search in sub-attributes and return matching attribute types', () => {
```
Comment (Zorin95670) [inline tests/unit/models/Component.spec.js:630]: Add check on object type

File: tests/unit/models/ComponentAttribute.spec.js (hunk 11 above)
Comment (justinehell) [reply tests/unit/models/ComponentAttribute.spec.js:144]: done

File: tests/unit/models/ComponentAttribute.spec.js [hunk 15]
```diff
@@ -101,6 +101,108 @@ describe('Test class: ComponentAttribute', () => {
...
+            type: 'Link',
+            rules: {
+              values: null,
+              min: 1,
+              max: 5,
+            },
+          },
+        });
+
+        componentAttribute.replaceLink('wrongId', 'newId');
```
[the remark this answers is not in the document]
Comment (justinehell) [reply tests/unit/models/ComponentAttribute.spec.js:164]: done

File: tests/unit/models/ComponentAttribute.spec.js (hunk 12 above)
Comment (justinehell) [reply tests/unit/models/ComponentAttribute.spec.js]: done

File: src/models/Component.js (hunk 13 above)
Comment (justinehell) [reply src/models/Component.js]: done

File: tests/unit/models/Component.spec.js (hunk 14 above)
Comment (justinehell) [reply tests/unit/models/Component.spec.js]: done

**Review by Zorin95670 — APPROVED** [review]

**Review by k-y0u — APPROVED** [review]

**Pull request merged**